### PR TITLE
gui: Add Open External Wallet action

### PR DESCRIPTION
**Pull request by promag**
This PR adds the ability to open external wallets on the GUI.

Screenshot 2019-09-08 at 18 14 36
Screenshot 2019-09-08 at 18 13 17

### CODE REVIEW

**Comment by Sjors** [comment]
Concept ACK, but let's put this in the Open Wallet sub menu. You could put a separator underneath the list of known wallets and then call the menu item "Open file...". I don't like the word "External" because it could get confusing if we add (e.g.) hardware wallet support.

**Comment by promag** [comment]
@Sjors yes that was already brought up by @jnewbery and in IRC I suggested "File -> Open Wallet -> Other..." On the end of the menu.

**Comment by promag** [comment]
@Sjors how about:
screenshot 2019-01-19 at 17 26 36

**Comment by jnewbery** [comment]
> @Sjors how about:

Yes, this looks ideal to me

**Comment by promag** [comment]
Updated.

**Comment by promag** [comment]
Updated, still rebased on #15195.

**Comment by Sjors** [comment]
tACK b537d29, but note that you can only open directory based wallets

**Comment by promag** [comment]
@Sjors right, and it looks like a pain to try to open both with Qt 😕

**Comment by promag** [comment]
Is this for 0.18?

**Comment by Sjors** [comment]
Needs rebase because Open Wallet got merged. That shouldn't impact my earlier tACK, so we might be able to add this to 0.18 today.

**Comment by jonasschnelli** [comment]
Needs rebase

**Comment by fanquake** [comment]
@promag Given that this is tagged for `0.19.0`, can you rebase?

**Comment by promag** [comment]
Sure, in a couple of hours.

**Comment by promag** [comment]
Rebased, and refactored. Reverted to add the action in the File menu, not in the wallets menu, so at the moment it's like:

Screenshot 2019-09-08 at 12 57 41

Previous version:

screenshot 2019-01-19 at 17 26 36

@jonasschnelli @Sjors @fanquake  Let me know your thoughts.

I'll update OP after this decision.

**Comment by fanquake** [comment]
@promag I prefer the previous version, with the action in the wallet menu. Will let other weigh in.

**Comment by promag** [comment]
You mean the open menu with:
 - no wallets available
 - separator
 - other...

**Comment by fanquake** [comment]
The [one currently in the PR description](https://user-images.githubusercontent.com/3534524/51430274-bc8e2f00-1c10-11e9-82b9-c736d5d43ef6.png) and [in this comment](https://github.com/bitcoin/bitcoin/pull/15204#issuecomment-455799571).

**Comment by jnewbery** [comment]
I agree with @fanquake that keeping it in the 'Open Wallet' submenu makes more sense.

**Comment by promag** [comment]
Thanks, also updated OP.

**Comment by meshcollider** [comment]
Concept ACK

**Comment by jonatack** [comment]
Concept ACK, will review this weekend.

**Comment by instagibbs** [comment]
got this message on stderr on qt startup with this PR:
`Gtk-Message: 16:07:25.804: GtkDialog mapped without a transient parent. This is discouraged.`

I tried opening up a random wallet filed I had, but it was greyed out in the file selection? Am I missing something?

**Comment by promag** [comment]
@instaggibbs see https://github.com/bitcoin/bitcoin/pull/15204#issuecomment-463132136

**Comment by instagibbs** [comment]
ok that's going to have to be communicated in some obvious way...

**Comment by instagibbs** [comment]
what does `directory based wallets` even mean? how is this detected?

**Comment by ryanofsky** [comment]
> what does `directory based wallets` even mean i.e., how is this detected?

The underlying wallet code can only open:

1) Wallet files named wallet.dat
2) Wallet files with names other than `wallet.dat` that are located directly in the top level wallet_dir

It hasn't been possible since #11687 to even create wallet files named anything other than `wallet.dat` so the second option just exists for backwards compatibility.

Only creating wallet files named `wallet.dat` and referring to wallets by directory rather than file name was done to address concerns from luke and others https://github.com/bitcoin/bitcoin/pull/11687#issuecomment-344381289 that if users only backed up individual wallet files instead of whole wallet directories they could lose data.

**Comment by ryanofsky** [comment]
Haven't tested this but looking at the implementation it seems like this might be lacking the error checking from the `loadwallet` RPC that ensures that the wallet which the user is trying to open actually exists, and that CreateWalletFromFIle code will not go and create at new empty wallet at the specified location:

https://github.com/bitcoin/bitcoin/blob/fb4f5beb6ede4aadeaff779cd67a0f6665419488/src/wallet/rpcwallet.cpp#L2577-L2585

Could be a pretty scary user experience to try to load an existing wallet but choose the wrong path, and see a new completely empty wallet instead of a "location does not contain wallet.dat" error.

If adding the check is necessary, I'd suggest moving the code out of the loadwallet RPC to a common location instead of adding duplicate code for the gui.

**Comment by promag** [comment]
> Could be a pretty scary user experience to try to load an existing wallet but choose the wrong path, and see a new completely empty wallet instead of a "location does not contain wallet.dat" error.

@ryanofsky indeed, currently it's like "load or create" :trollface: will change accordingly.

> ok that's going to have to be communicated in some obvious way...

@instagibbs you mean in the release notes or do you have something else in mind?

**Comment by instagibbs** [comment]
oh.... you have to click "open" while in the right directory or after clicking once on that directory holding the wallet file?

that's not even close to intuitive :grimacing:

**Comment by instagibbs** [comment]
So this either needs an extensive explanation before the user ventures to use it, or it should allow the user to click and open a wallet.dat in a directory, in addition to the directory itself. Latter seems to make most sense?

**Comment by promag** [comment]
@instagibbs mind testing again? Now you can select the wallet.dat or the parent directory.

@ryanofsky now it doesn't create the wallet, thanks!

**Comment by Sjors** [comment]
On macOS, when I select a directory ~/temp/A it complains:

Schermafbeelding 2019-09-16 om 10 45 08

When I double click on the directory A (so it shows the wallet.dat file, etc) it still gives the above error.

It only works when I select the wallet.dat file.

**Comment by MarcoFalke** [comment]
This has missed the deadline in #15940

**Comment by jnewbery** [comment]
I don't think the wallet name should display the path to that wallet:

Screenshot from 2019-09-16 11-54-07
Screenshot from 2019-09-16 11-54-27

**Comment by ryanofsky** [comment]
> So this either needs an extensive explanation before the user ventures to use it, or it should allow the user to click and open a wallet.dat in a directory, in addition to the directory itself. Latter seems to make most sense?

Allowing opening of wallet.dat files seems good. Would just need to strip the last `wallet.dat` component to make the path a wallet location. Another option would be to use a directory selector instead of file selector. https://stackoverflow.com/questions/9618381/how-to-specify-the-qfiledialoggetexistingdirectory-method

> I don't think the wallet name should display the path to that wallet:

I kind of do, but anything in particular you'd like to replace it with? It should be ok to replace it with something something shorter in the GUI selector. I'd be more wary of changing the output of `getwalletinfo` RPC, or changing the invariant that whatever string you pass to `-wallet` or `loadwallet` or `createwallet` will uniquely identify the wallet while it is open, and be the wallet endpoint and `-rpcwallet` argument you pass to bitcoin-cli.

**Comment by promag** [comment]
> Would just need to strip the last wallet.dat component to make the path a wallet location 

That's done here: https://github.com/bitcoin/bitcoin/pull/15204/commits/9e1fdd7f94161ecfbe7a4247fb2a672d5190ff83#diff-dc8c2957962a3032071c55c70e120932R331 but apparently shouldn't be done for directories.

> I don't think the wallet name should display the path to that wallet:

@jnewbery this wasn't changed here, it's already the case if you use `loadwallet` with a path. Let's move that discussion elsewhere? 

@Sjors I think I've fixed it, mind checking?

**Comment by promag** [comment]
BTW I'm avoiding `QFileDialog` static methods, such as [QFileDialog::getOpenFileName](https://doc.qt.io/qt-5/qfiledialog.html#getOpenFileName), which have the following drawback (from the doc):
> Warning: Do not delete parent during the execution of the dialog. If you want to do this, you should create the dialog yourself using one of the QFileDialog constructors.

#15310 is related to this warning.

**Comment by Sjors** [comment]
In 91d6e0b I'm able to to open a wallet by just selecting the directory. Opening the directory and opening the wallet.dat file work as well.

Agree with @jnewbery about using the wallet name instead of the full path. Since the RPC has the same problem, we can deal with that in a followup.

**Comment by instagibbs** [comment]
ACK on punting the full path issue until a followup

**Comment by instagibbs** [comment]
Now when I use it, I can only open the wallet by selecting a file within a directory with a `wallet.dat` present. It lets me "open" any file in that directory, but it opens up the specific `wallet.dat` that is in the same directory, I think.

**Comment by promag** [comment]
I'm tempted to allow selecting either wallet.dat or a directory.

**Comment by ryanofsky** [comment]
> I'm tempted to allow selecting either wallet.dat or a directory.

If I were implementing this, I'd use a folder selector (https://stackoverflow.com/questions/13299283/folder-browser-dialog-in-qt ), not a file selector, and only allow choosing directories containing wallet.dat files (disable the choose / select button for other directories).

But if there's a reason to list files as well as directories, I'd think there should be a way to filter the file list to only show wallet.dat files.

**Comment by instagibbs** [comment]
> I'd use a folder selector

concept ACK on this suggestion. I don't think I would have been overly confused with this originally.

**Comment by promag** [comment]
Done, now can only select folders.

**Comment by instagibbs** [comment]
ok, this is much more natural. quick tACK https://github.com/bitcoin/bitcoin/pull/15204/commits/4e8cb3fd5efe2d108d8cf4017d197fb3e7c8eb40

**Comment by Sjors** [comment]
Agree the behavior in 4e8cb3f is nicer. I haven't reviewed the code.

**Review by jonatack — COMMENTED** [review]
Light code review and quick manual test ACK  4e8cb3fd5efe2d108d8cf4017d197fb3e7c8eb40 apart from my comments below and a bit more manual testing -- I may have encountered weird behavior as described in my comments below. I agree with @jnewbery on the full path behavior and  understand @ryanofsky's points, and don't have a proposal yet that is more concrete than having a way to see the full path that is discreet and user-friendly. Punting for a follow-up duly noted.

**Comment by jonatack** [comment]
> I don't like the word "External" because it could get confusing if we add (e.g.) hardware wallet support.

I'm hesitant as well with the ambiguity of the word "external" -- should we be more precise?

**Comment by instagibbs** [comment]
How about "wallet in external directory" and variations thereof

**Comment by jonatack** [comment]
Yes, "external directory" or "in another directory" seems already better (unless the plan is to include external drives and HWs in the same category later on)

**Comment by ryanofsky** [comment]
> How about "wallet in external directory" and variations thereof

I like the "Open a wallet in an external directory" Jon suggested

**Comment by michaelfolkson** [comment]
The multiple wallets should be in separate directories by default? In which case why not just "Other wallet" or "Additional wallet"? I don't think attention should be brought to it being in an external directory if that is the default.

**Comment by ryanofsky** [comment]
> The multiple wallets should be in separate directories by default? In which case why not just "Other wallet" or "Additional wallet"? I don't think attention should be brought to it being in an external directory if that is the default.

External means outside the bitcoin wallet dir (`~/.bitcoin/wallets` on linux) where the list of wallets comes from. I like your "Open other wallet..." suggestion, though. Maybe saying "external" just adds confusion.

**Comment by jonatack** [comment]
Maybe "Open a wallet located in a different directory"

File: src/qt/bitcoingui.cpp [hunk 1]
```diff
@@ -337,6 +337,9 @@ void BitcoinGUI::createActions()
     m_open_wallet_action->setStatusTip(tr("Open a wallet"));
     m_open_wallet_menu = new QMenu(this);
 
+    m_open_external_wallet_action = new QAction(tr("Other..."), this);
+    m_open_external_wallet_action->setStatusTip(tr("Open a wallet on external location"));
```
[the remark this answers is not in the document]
Comment (promag) [reply src/qt/bitcoingui.cpp]: Picked "Open a wallet in an external directory".

File: src/qt/walletcontroller.cpp [hunk 2]
```diff
@@ -281,10 +283,63 @@ void OpenWalletActivity::open(const std::string& path)
...
+
+    if (m_wallet_model) Q_EMIT opened(m_wallet_model);
+
+    Q_EMIT finished();
+}
+
+void OpenExternalWalletActivity::open()
+{
+    Q_ASSERT(!m_file_dialog);
+    m_file_dialog = new QFileDialog(m_parent_widget, tr("Open External wallet"), QDir::homePath());
```
[the remark this answers is not in the document]
Comment (promag) [reply src/qt/walletcontroller.cpp]: Changed to "Open Wallet", I think it's enough.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -157,9 +157,31 @@ std::shared_ptr<CWallet> LoadWallet(interfaces::Chain& chain, const WalletLocati
...
+        return nullptr;
+    }
+
+    if (fs::is_directory(location.GetPath())) {
+        // The given filename is a directory. Check that there's a wallet.dat file.
+        fs::path wallet_dat_file = location.GetPath() / "wallet.dat";
+        if (fs::symlink_status(wallet_dat_file).type() == fs::file_not_found) {
+            exists = false;
+            error = "Directory " + location.GetName() + " does not contain a wallet.dat file.";
+            return nullptr;
```
[the remark this answers is not in the document]
Comment (promag) [reply src/wallet/wallet.cpp]: Can't reproduce, please let me know if you hit it again.

**Comment by jonatack** [comment]
Will test more and re-review. Some edge case test ideas from today's PR review meeting:
- no wallet in directory
- multiple wallets in directory
- a non-wallet file named wallet.dat
- deleting the wallet file while trying to load it

**Comment by promag** [comment]
> Will test more and re-review. Some edge case test ideas from today's PR review meeting:
> 
> * no wallet in directory

it already fails:

Screenshot 2019-09-18 at 22 50 05


> * multiple wallets in directory

not supported - it's already the case with RPC `loadwallet`:

```
bitcoin-cli -regtest loadwallet /Users/joao/Desktop/test.dat

Wallet file verification failed: Invalid -wallet path '/Users/joao/Desktop/test.dat'. -wallet path should point to a directory where wallet.dat and database/log.?????????? files can be stored, a location where such a directory could be created, or (for backwards compatibility) the name of an existing data file in -walletdir ("/Users/joao/Library/Application Support/Bitcoin/regtest/wallets") (code -4)
```

> * a non-wallet file named wallet.dat

not supported ☝️ 

> * deleting the wallet file while trying to load it

Didn't test this here, and not sure how to reliable test it.

**Comment by fanquake** [comment]
re-Concept ACK. Here's the dialog on macOS:

open_dialog

A few thoughts:

It'd be good if the GUI realised when we opened a wallet that's actually inside our wallet directory via the `Other...` menu, otherwise you end up with the wallet loaded, but not greyed out:
open

This seems to have created at least one path to a segfault, where if you load duplicate (copy of a) wallets we don't catch the BerkeleyBatch exception. I don't think it was possible to do this via the GUI before (it's also currently caught when loading wallets via the RPC etc).
```
libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: BerkeleyBatch: Can't open database wallet.dat (duplicates fileid ccd6a006040000010ccf2f5cc535010000000000 from wallet.dat)

2019-09-19T03:02:31Z GUI: Qt has caught an exception thrown from an event handler. Throwing
exceptions from an event handler is not supported in Qt.
You must not let any exception whatsoever propagate through Qt code.
If that is not possible, in Qt 5 you must at least reimplement
QCoreApplication::notify() and catch all exceptions there.
```

> Maybe saying "external" just adds confusion.

Agree.

Noticed at least one quirk where Qt fails to display file paths correctly:

path_confusion

**Comment by promag** [comment]
> It'd be good if the GUI realised when we opened a wallet that's actually inside our wallet directory via the `Other...` menu, otherwise you end up with the wallet loaded, but not greyed out:

Yes, different PR I guess.

**Comment by jonatack** [comment]
@promag Thanks for the update. While testing f60f80e599284b4f098b54766e2221ebcd11f587, I reproduced the [issue mentioned above](https://github.com/bitcoin/bitcoin/pull/15204/#discussion_r325757097) on  Debian 4.19.37-5+deb10u2 (2019-08-08) x86_64 GNU/Linux. Opening a folder causes the GUI display the error, then a second later attempts to open the wallet anyway and hangs in this state.

Screenshot from 2019-09-19 12-26-00

**Comment by jonatack** [comment]
Copying a wallet folder elsewhere with a different name, and attempting to load it in the GUI:
```
$ qt/bitcoin-qt -testnet
terminate called after throwing an instance of 'std::runtime_error'
  what():  BerkeleyBatch: Can't open database wallet.dat (duplicates fileid c4003e0100fe0000fd9fb0c2a56b000000000000 from wallet.dat)
Aborted
```

Screenshot of a wallet successfully loaded after I moved it from ~/.bitcoin to ~/ with an ampersand (&) displayed correctly:

Screenshot from 2019-09-19 12-41-10

**Comment by promag** [comment]
@jonatack regarding the "duplicates fileid " crash see https://github.com/bitcoin/bitcoin/issues/16776#issuecomment-528672904

**Comment by promag** [comment]
@jonatack just to be sure, if you repeat the above tests with `loadwallet` in the console you would get the same problems?

**Comment by jonasschnelli** [comment]
Tested a bit.
What I think is hard to understand is that one needs to select a folder rather than a wallet.dat file.

I think either we allow to select wallet.dat files or inform the user that only wallet-directories are allowed ("select a wallet folder").

**Comment by promag** [comment]
> select a wallet folder

I like that.

**Comment by Sjors** [comment]
If you make `OpenExternalWalletActivity` a subclass of `OpenWalletActivity` it would save duplicate code, which is especially nice in `Open[External]WalletActivity::open()` with its `QTimer::singleShot`.

**Comment by jonasschnelli** [comment]
Needs addressing comments or should otherwise be closed.

**Comment by promag** [comment]
Will rebase and open in gui repo.

**Comment by promag** [comment]
> Tested a bit.
> What I think is hard to understand is that one needs to select a folder rather than a wallet.dat file.
> 
> I think either we allow to select wallet.dat files or inform the user that only wallet-directories are allowed ("select a wallet folder").

I've tried to support selecting files and folders but this kind of sucks because in order be intuitive it needs non native file dialog - it allows to set a proxy model.

Anyway, I think the best approach, and considering sqlite wallets, selecting wallet.dat is indeed the best option for the moment.

**Pull request closed without merging**